js/state/syncUtil.js

```diff
@@ -280,12 +300,37 @@ module.exports.newObjectId = (objectPath) => {
 }
 
 /**
+ * Given a bookmark folder's folderId, get or set its object ID.
+ * @param {number} folderId
+ * @returns {Array.<number>}
+ */
+module.exports.findOrCreateFolderObjectId = (folderId) => {
+  if (!folderId) { return undefined }
+  const AppStore = require('../stores/appStore')
+  const appState = AppStore.getState()
+  const folderEntry = appState.get('sites').findEntry((site, _index) => {
+    return site.get('folderId') === folderId
+  })
+  if (!folderEntry) { return undefined }
+  const folderIndex = folderEntry[0]
+  const folder = folderEntry[1]
+  const objectId = folder.get('objectId')
+  if (objectId) {
+    return objectId.toJS()
+  } else {
+    return module.exports.newObjectId(['sites', folderIndex])
+  }
+}
+
+/**
  * Converts a site object into input for sendSyncRecords
  * @param {Object} site
  * @param {number=} siteIndex
+ * @param {boolean} generateObjectIds objectIds are generated via AppActions IPC
+ *   which is async. You may prefer to generate it for immediate use.
```

Comment:
thx, i feel it now and removed the param.

the problem i had was in going from `parentFolderId` to `parentFolderObjectId`.

example:
- 2 bookmarks in a folder
- sync bookmark 1. it looks up its folder by parentFolderId, and sees the folder has no objectId so sets it.
- sync bookmark 2. it also looks up folder by parentFolderId, and again sees the folder has no objectId so sets it. this happens faster than the setting of the objectId.

i've retained the memoing in sync.js, and updated createSiteData() to use `site.parentFolderObjectId` if already set.